poetry/repositories/installed_repository.py

```diff
@@ -99,6 +101,127 @@ def is_vcs_package(cls, package: Union[Path, Package], env: Env) -> bool:
         else:
             return True
 
+    @classmethod
+    def create_package_from_distribution(
+        cls, distribution: "Distribution", env: "Env"
+    ) -> Package:
+        # We first check for a direct_url.json file to determine
+        # the type of package.
+        path = Path(str(distribution._path))
+
+        if (
+            path.name.endswith(".dist-info")
+            and path.joinpath("direct_url.json").exists()
+        ):
+            return cls.create_package_from_pep610(distribution)
+
+        is_standard_package = env.is_path_relative_to_lib(path)
+
+        source_type = None
+        source_url = None
+        source_reference = None
+        source_resolved_reference = None
+        if is_standard_package:
+            if path.name.endswith(".dist-info"):
+                paths = cls.get_package_paths(
+                    env=env, name=distribution.metadata["name"]
+                )
+                if paths:
+                    is_editable_package = False
+                    for src in paths:
+                        if cls.is_vcs_package(src, env):
+                            (
+                                source_type,
+                                source_url,
+                                source_reference,
+                            ) = cls.get_package_vcs_properties_from_path(src)
+                            break
+
+                        if not (
+                            is_editable_package or env.is_path_relative_to_lib(src)
+                        ):
+                            is_editable_package = True
+                    else:
+                        # TODO: handle multiple source directories?
+                        if is_editable_package:
+                            source_type = "directory"
+                            source_url = paths.pop().as_posix()
+        else:
+            if cls.is_vcs_package(path, env):
+                (
+                    source_type,
+                    source_url,
+                    source_reference,
+                ) = cls.get_package_vcs_properties_from_path(
+                    env.path / "src" / canonicalize_name(distribution.metadata["name"])
+                )
+            else:
+                # If not, it's a path dependency
+                source_type = "directory"
+                source_url = str(path.parent)
+
+        package = Package(
+            distribution.metadata["name"],
+            distribution.metadata["version"],
+            source_type=source_type,
+            source_url=source_url,
+            source_reference=source_reference,
+            source_resolved_reference=source_resolved_reference,
+        )
+        package.description = distribution.metadata.get("summary", "")
+
+        return package
+
+    @classmethod
+    def create_package_from_pep610(cls, distribution: "Distribution") -> Package:
```

Comment:
On a general note (not a blocker for this PR), I feel that we have various ways to "create a package" spread across the code base, we should at some point bring them all together so it becomes slightly easier to maintain.

Reply:
For now, I don't anticipate that we will use it elsewhere. If we ever do we will see where we can put it, I just don't want us to make early refactoring if we don't need to.